### PR TITLE
Wasi3 review

### PR DESCRIPTION
**Pull request by andreaTP**
I think I have tackled all the outstanding points, great work @bhelx !

I have attempted to re-compile everything, unfortunately, I'm unable to reproduce the output of `greet-wasi.rs` and `hello-wasi.rs` (seems like the exported `_start` is not generated), but we can file an issue and get back to it later on.

### CODE REVIEW

**Review by bhelx — APPROVED** [review]
Looks good, thanks for picking this up :)

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -4,7 +4,6 @@ on:
   push:
     branches: [ main ]
   pull_request:
-    branches: [ main ]
```
Comment (bhelx) [inline .github/workflows/ci.yaml:7]: Is this meant to be removed?

Comment (andreaTP) [reply .github/workflows/ci.yaml:7]: yes, it doesn't hurt, it will run the CI on any PR even not targeting `main`

Comment (bhelx) [reply .github/workflows/ci.yaml:7]: oh i see, makes sense

**Pull request merged**